src/parser/parser.ts

```diff
@@ -0,0 +1,167 @@
+import forEach from "lodash/forEach";
+
+import {
+  LinkDiscriminator,
+  SafeLink,
+  Link,
+  LinkContent,
+  Arguments,
+} from "./types";
+
+/*
+ * parseLink takes a striped matrix.to hash link (without the '#/' prefix)
+ * and parses into a Link. If the parse failed the result will
+ * be ParseFailed
+ */
+export function parseHash(hash: string): Link {
+  const [identifier, args] = hash.split("?");
+
+  const kind = identifyTypeFromRegex(
+    identifier,
+    verifiers,
+    LinkDiscriminator.ParseFailed
+  );
+
+  let parsedLink: LinkContent = {
+    identifier,
+    arguments: parseArgs(args),
+    originalLink: hash,
+  };
+
+  if (kind === LinkDiscriminator.Permalink) {
+    const { roomKind, roomLink, eventId } = parsePermalink(identifier);
+
+    return {
+      kind,
+      ...parsedLink,
+      roomKind,
+      roomLink,
+      eventId,
+    };
+  }
+
+  return {
+    kind,
+    ...parsedLink,
+  };
+}
+
+/*
+ * Parses a permalink.
+ * Assumes the permalink is correct.
+ */
+export function parsePermalink(identifier: string) {
+  const [roomLink, eventId] = identifier.split("/");
+  const roomKind = identifyTypeFromRegex(
+    roomLink,
+    roomVerifiers,
+    // This is hacky but we're assuming identifier is a valid permalink
+    LinkDiscriminator.Alias
+  );
+
+  return {
+    roomKind,
+    roomLink,
+    eventId,
+  };
+}
+
+/*
+ * parseArgs parses the <extra args> part of matrix.to links
+ */
+export function parseArgs(args: string): Arguments {
+  const params = new URLSearchParams(args);
+  const _federated = params.get("federated");
+  const federated = _federated !== null ? _federated === "true" : null;
```

Comment:
Do expect to need `federated` to implement the designs? It's likely this will be removed from the MSC and replaced with something like `hs` stating a required HS to use, but it's not quite clear yet.

If you do not need it for the designs, I would suggest removing it here. If you do need, then let's keep it for now, and it can reworked later once the MSC is updated.

Reply:
It's not needed for the designs so I will remove it.